docs/architecture/adr-110-remote-mempool.md

```diff
@@ -0,0 +1,364 @@
+    //    where the consensus engine may have failed and restarted.
+    //
+    // The consensus engine will only ever make one call to TxsAvailable and
+    // will attempt to keep the connection alive.
+    rpc TxsAvailable(TxsAvailableRequest) returns (stream TxsAvailableResponse);
+}
+
+message FetchRequest {
+    // Fetch as many transactions as possible that cumulatively take up at most
+    // this number of bytes. Setting this value to -1 implies no limit (this
+    // implies that the remote mempool is fully in control of this value).
+    int64 max_bytes = 1;
+}
+
+message FetchResponse {
+    // Transactions to be included in the proposal.
+    repeated bytes txs = 1;
+}
+
+message ReapRequest {
+    // The current height of the chain.
+    uint64 height = 1;
+
+    // A list of IDs of transactions to be removed from the mempool. At present
+    // this is a list of SHA256 hashes of the transactions.
+    repeated bytes tx_ids = 2;
+}
+
+message ReapResponse {}
+
+message TxsAvailableRequest {}
+
+// TxsAvailableResponse is to be sent once upon initiation of the TxsAvailable
+// request, as well as once for each new height, iff new transactions are
+// available to be fetched by the consensus engine.
+message TxsAvailableResponse {
+    // Supplied for informational purposes to the consensus engine to indicate
+    // the height seen by the mempool during the last call to Reap.
+    uint64 last_reap_height = 1;
+}
+```
+
+**Notes**:
+
+- The terminology used in the gRPC interface is different to that used in the
+  [`Mempool`] interface. The term "reap" technically implies removal from the
+  mempool, but the [`Mempool`] interface incorrectly uses this term to imply
+  fetching a batch of transactions.
+
+- The gRPC interface purposefully does not facilitate limiting fetched
+  transactions by gas in an effort to separate consensus- and
+  application-related concerns (as per [RFC 011]). Should remote mempool
+  developers want to limit returned transactions by gas, this should be
+  implemented as part of the configuration of the remote mempool.
+
+### Impact and properties
+
+A new mempool that implements the [`Mempool`] interface will be created, and
+when enabled it will have the following impacts/properties:
+
+1. The mempool-related ABCI methods will not be called.
+2. The following methods will do nothing:
+   - `CheckTx`, since transaction validation is assumed to happen prior to
+     transaction insertion in the mempool.
+   - `FlushAppConn`, which is only relevant for use via ABCI.
+   - `RemoveTxByKey`, which is only used in the callback in the context of ABCI.
+   - `ReapMaxTxs`, which is only used internally and by the `unconfirmed_txs`
+     RPC endpoint.
+   - `Flush`, which is only ever used via the `unsafe_flush_mempool` RPC
+     endpoint.
+   - `Size` and `SizeBytes`, since these are only used for informational
+     purposes in the RPC.
```

Comment:
How will operators know the status of the mempool? Should this become then the application's concern to provide or should we support some of these endpoints and forward them to the app? If they are not implemented we return empty responses.

Reply:
I agree that this sentence is not very clear. What I understand is that txs are gossiped through the mempool network. And that txs received from (mempool) peers need to be validated by the application. Comet is not involved at all here. And not necessarily a mempool instance is running on the same machine as a Comet instance.

Reply:
It's up to the remote mempool developer to take care of that. It's a separate process that needs to be monitored independently. Us taking on that responsibility in any way starts to inappropriately mix concerns.